sharktank/sharktank/models/t5/t5.py

```diff
@@ -0,0 +1,1067 @@
+                    d_kv=d_kv,
+                    num_heads=num_heads,
+                    activation_dtype=activation_dtype,
+                    layer_norm_epsilon=layer_norm_epsilon,
+                )
+            )
+
+        self.layer.append(
+            T5LayerFF(
+                theta=theta,
+                is_gated_act=is_gated_act,
+                dense_act_fn=dense_act_fn,
+                layer_norm_epsilon=layer_norm_epsilon,
+                activation_dtype=activation_dtype,
+            )
+        )
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask=None,
+        position_bias=None,
+        encoder_hidden_states=None,
+        encoder_attention_mask=None,
+        encoder_decoder_position_bias=None,
+        layer_head_mask=None,
+        cross_attn_layer_head_mask=None,
+        past_key_value=None,
+        use_cache=False,
+        output_attentions=False,
+        return_dict=True,
+    ):
+        if past_key_value is not None:
+            if not self.is_decoder:
+                logger.warning(
+                    "`past_key_values` is passed to the encoder. Please make sure this is intended."
+                )
+            expected_num_past_key_values = 2 if encoder_hidden_states is None else 4
+
+            if len(past_key_value) != expected_num_past_key_values:
+                raise ValueError(
+                    f"There should be {expected_num_past_key_values} past states. "
+                    f"{'2 (key / value) for cross attention. ' if expected_num_past_key_values == 4 else ''}"
+                    f"Got {len(past_key_value)} past key / value states"
+                )
+
+            self_attn_past_key_value = past_key_value[:2]
+            cross_attn_past_key_value = past_key_value[2:]
+        else:
+            self_attn_past_key_value, cross_attn_past_key_value = None, None
+
+        self_attention_outputs = self.layer[0](
+            hidden_states,
+            attention_mask=attention_mask,
+            position_bias=position_bias,
+            layer_head_mask=layer_head_mask,
+            past_key_value=self_attn_past_key_value,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+        )
+        hidden_states, present_key_value_state = self_attention_outputs[:2]
+        attention_outputs = self_attention_outputs[
+            2:
+        ]  # Keep self-attention outputs and relative position weights
+
+        # clamp inf values to enable fp16 training
+        if hidden_states.dtype == torch.float16:
+            clamp_value = ops.elementwise(
+                torch.where,
+                ops.elementwise(torch.isinf, hidden_states).any(),
+                torch.finfo(hidden_states.dtype).max - 1000,
+                torch.finfo(hidden_states.dtype).max,
+            )
+            hidden_states = ops.elementwise(
+                torch.clamp, hidden_states, min=-clamp_value, max=clamp_value
+            )
```

Comment:
@archana-ramalingam, you mentioned some training remnants, do you mean these?
I am a bit reluctant to remove these as if the train was done with clamping, shouldn't inference also clamp? In general we want to run the same function as it was trained on.
The question is if we get a trained fp16 T5, would it have used this clamping during training?

Reply:
Yes, this was it. Are you able to test if removing this block made a diff in the results?

Reply:
I have not done any float16 tests yet. Only float32 as this is the quantization of the original model from Google. I think this is an addition others made when doing fine tunning. I want to leave it in as this is also how inference works on the HF model.
This infinity probably comes from some previous calculation and if it appears it will blow up downstream results. It probably happens rarely.